simple/multi_mr.c

```diff
@@ -0,0 +1,496 @@
+	if (ret < 0)
+		return ret;
+
+	*cntr_next += 1;
+
+	return 0;
+}
+
+static int free_mr_res()
+{
+	int ret, i;
+
+	if (g_test_state.mr_state) {
+		for (i = 0; i < g_test_state.mr_count; i++) {
+			ret = fi_close(&g_test_state.mr_state[i].mr->fid);
+			if (ret) {
+				FT_PRINTERR("fi_close", ret);
+				return ret;
+			}
+
+			ret = fi_close(&g_test_state.mr_state[i].rcntr->fid);
+			if (ret) {
+				FT_PRINTERR("fi_close", ret);
+				return ret;
+			}
+
+			if (g_test_state.mr_state[i].buf)
+				free(g_test_state.mr_state[i].buf);
+		}
+		free(g_test_state.mr_state);
+	}
+
+	return 0;
+}
+
+static int alloc_multi_mr_res()
+{
+	int ret = 0, i;
+	struct fi_cntr_attr cntr_attr = {0};
+	cntr_attr.events = FI_CNTR_EVENTS_COMP;
+	cntr_attr.wait_obj = FI_WAIT_UNSPEC;
+
+	if ((fi->domain_attr->mr_key_size * 8) < g_test_state.mr_count) {
+		fprintf(stderr, "ERROR: too many memory regions for unique mr keys");
+		return -1;
+	}
+
+	g_test_state.mr_count = 1ULL << g_test_state.mr_count;
+
+	g_test_state.mr_state =
+		malloc(sizeof(struct test_mr_state) * g_test_state.mr_count);
+	if (!g_test_state.mr_state)
+		return -errno;
+
+	for (i = 0; i < g_test_state.mr_count; i++) {
+		g_test_state.mr_state[i].remote = malloc(sizeof(struct fi_rma_iov));
+		if (!g_test_state.mr_state[i].remote)
+			return -errno;
+
+		ret = fi_cntr_open(domain,
+				&cntr_attr,
+				&g_test_state.mr_state[i].rcntr,
+				NULL);
+		if (ret) {
+			FT_PRINTERR("fi_cntr_open", ret);
+			return ret;
+		}
+
+		g_test_state.mr_state[i].buf = malloc(opts.transfer_size);
+		if (!g_test_state.mr_state[i].buf)
+			return -errno;
+
+		g_test_state.mr_state[i].remote->len = opts.transfer_size;
+		g_test_state.mr_state[i].remote->addr = 0;
+
+		memset(g_test_state.mr_state[i].buf, 0, opts.transfer_size);
+
+		g_test_state.mr_state[i].rcntr_next = 1;
+
+		memset(&g_test_state.mr_state[i].remote->key, 0, sizeof(uint64_t));
+
+		g_test_state.mr_state[i].remote->key = i + 1;
+		ret = fi_mr_reg(domain,
+				g_test_state.mr_state[i].buf,
+				opts.transfer_size,
+				FI_REMOTE_WRITE,
+				0,
+				g_test_state.mr_state[i].remote->key,
+				0,
+				&g_test_state.mr_state[i].mr,
+				NULL);
+		if (ret) {
+			FT_PRINTERR("fi_mr_reg", ret);
+			return ret;
+		}
+
+		printf("fi_mr_reg(fi_domain = %p, buffer = %p, memory_region_size = %d, &memory_region = %p, mr_key %lu)\n",
+				domain,
+				g_test_state.mr_state[i].buf,
+				opts.transfer_size,
+				&g_test_state.mr_state[i].mr,
+				g_test_state.mr_state[i].remote->key);
+
+		ret = fi_mr_bind(g_test_state.mr_state[i].mr,
+				&g_test_state.mr_state[i].rcntr->fid,
+				FI_REMOTE_WRITE);
+		if (ret) {
+			FT_PRINTERR("fi_mr_bind", ret);
+			return ret;
+		}
+	}
+
+	return ret;
+}
+
+static int check_buffer(uint8_t *buffer, size_t length, uint8_t pattern)
+{
+	int i;
+
+	for (i = 0; i < length; i++) {
+		if (buffer[i] != pattern) {
+			fprintf(stdout, "ERROR: failed check, %d != %d\n", buffer[i], pattern);
+			return -1;
+		}
+	}
+
+	return 0;
+}
```

Comment:
There are existign calls to set/check data buffers.  Please use or adapt those.  See ft_check_buf()/ft_fill_buf()